core/chains/evm/chain.go

```diff
@@ -0,0 +1,258 @@
+	"github.com/smartcontractkit/chainlink/core/utils"
+	"go.uber.org/multierr"
+	"go.uber.org/zap"
+	"gorm.io/gorm"
+)
+
+type ChainIdentification interface {
+	IsL2() bool
+	IsArbitrum() bool
+	IsOptimism() bool
+}
+
+type Chain interface {
+	service.Service
+	ChainIdentification
+	Ready() error
+	ID() *big.Int
+	Client() eth.Client
+	Config() ChainScopedConfig
+	LogBroadcaster() log.Broadcaster
+	HeadBroadcaster() httypes.HeadBroadcaster
+	TxManager() bulletprooftxmanager.TxManager
+	HeadTracker() httypes.Tracker
+}
+
+var _ Chain = &chain{}
+
+type chain struct {
+	utils.StartStopOnce
+	chain           types.Chain
+	cfg             ChainScopedConfig
+	client          eth.Client
+	txm             bulletprooftxmanager.TxManager
+	logger          *zap.SugaredLogger
+	headBroadcaster httypes.HeadBroadcaster
+	headTracker     httypes.Tracker
+	logBroadcaster  log.Broadcaster
+	balanceMonitor  services.BalanceMonitor
+	keyStore        keystore.EthKeyStoreInterface
+}
+
+func newChain(dbchain types.Chain, globalLogger *logger.Logger, db *gorm.DB, gcfg config.GeneralConfig, keyStore keystore.EthKeyStoreInterface, advisoryLocker postgres.AdvisoryLocker, eventBroadcaster postgres.EventBroadcaster) (*chain, error) {
+	if gcfg.EthereumDisabled() {
+		return nil, errors.Errorf("cannot create new chain with ID %d, ethereum is disabled", dbchain.ID)
+	}
+	// TODO: Pass this logger into all subservices
+	cfg := newChainScopedConfig(db, gcfg, dbchain.ID.ToInt())
+	l := globalLogger.With("chainID", dbchain.ID.String())
+	serviceLogLevels, err := globalLogger.GetServiceLogLevels()
+	if err != nil {
+		return nil, err
+	}
+	client, err := newEthClientFromChain(dbchain)
+	if err != nil {
+		return nil, err
+	}
+	headTrackerLogger, err := globalLogger.InitServiceLevelLogger(logger.HeadTracker, serviceLogLevels[logger.HeadTracker])
+	if err != nil {
+		return nil, err
+	}
+	headBroadcaster := headtracker.NewHeadBroadcaster()
+	orm := headtracker.NewORM(db, *dbchain.ID.ToInt())
+	headTracker := headtracker.NewHeadTracker(headTrackerLogger, client, cfg, orm, headBroadcaster)
+	txm := bulletprooftxmanager.NewBulletproofTxManager(db, client, cfg, keyStore, advisoryLocker, eventBroadcaster)
+
+	// Highest seen head height is used as part of the start of LogBroadcaster backfill range
+	highestSeenHead, err2 := headTracker.HighestSeenHeadFromDB()
+	if err2 != nil {
+		return nil, err2
+	}
+
+	var balanceMonitor services.BalanceMonitor
+	if cfg.BalanceMonitorEnabled() {
+		balanceMonitor = services.NewBalanceMonitor(db, client, keyStore)
+	}
+
+	logBroadcaster := log.NewBroadcaster(log.NewORM(db), client, cfg, highestSeenHead)
+	// Log Broadcaster waits for other services' registrations
+	// until app.LogBroadcaster.DependentReady() call (see below)
+	logBroadcaster.AddDependents(1)
+
+	headBroadcaster.Subscribe(logBroadcaster)
+	headBroadcaster.Subscribe(txm)
+	headBroadcaster.Subscribe(balanceMonitor)
+
+	c := chain{
+		utils.StartStopOnce{},
```

Comment:
could you please define fields of this struct to have a more explicit way of populating these fields?

Reply:
Why so? Doing it this way allows us to make sure we didn't miss any fields.